### PR TITLE
Add SGNHT

### PR DESCRIPTION
**Pull request by SamDuffield**
Discussed in #62 

Adds SGNHT from [Ding et al](https://proceedings.neurips.cc/paper/2014/file/21fe5b8ba755eeaece7a450849876228-Paper.pdf) and also discussed in [Ma et al](https://arxiv.org/abs/1506.04696). Its essentially SGHMC with an adaptive friction coefficient.

### CODE REVIEW

File: posteriors/sgmcmc/sgnht.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+
+def build(
+    log_posterior: LogProbFn,
+    lr: float,
+    alpha: float = 0.01,
+    beta: float = 0.0,
+    temperature: float = 1.0,
+    momenta: TensorTree | float | None = None,
+    xi: float = None,
```
Comment (KaelanDt) [inline posteriors/sgmcmc/sgnht.py:21]: i would call this x0, more standard notation

File: tests/sgmcmc/test_sgnht.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    for _ in range(n_steps):
+        sgnht_state = sampler.update(sgnht_state, batch, inplace=False)
+
+        all_params = tree_map(
+            lambda x, y: torch.cat((x, y.unsqueeze(0))), all_params, sgnht_state.params
+        )
+
+        log_posts.append(sgnht_state.log_posterior.item())
+
+    burnin = 5000
```
Comment (KaelanDt) [inline tests/sgmcmc/test_sgnht.py:49]: is there an explanation for why burnin has to be larger than in the other tests?

**Review by KaelanDt — COMMENTED** [review]
Minor comments left

File: posteriors/sgmcmc/sgnht.py (hunk 1 above)
Comment (SamDuffield) [reply posteriors/sgmcmc/sgnht.py:21]: which greek letter is x0? 👀

File: tests/sgmcmc/test_sgnht.py (hunk 2 above)
Comment (SamDuffield) [reply tests/sgmcmc/test_sgnht.py:49]: Not really haha, tbh I think it should be larger for the other tests too

**Review by KaelanDt — APPROVED** [review]
lgtm

**Pull request merged**